junit-runtime/src/main/scala/com/novocode/junit/RichLogger.scala

```diff
@@ -0,0 +1,159 @@
+package com.novocode.junit
+
+import sbt.testing._
+
+import scala.collection.mutable
+
+import Ansi._
+
+final class RichLogger private(loggers: Array[Logger], settings: RunSettings) {
+
+  private[this] val currentTestClassName = new mutable.Stack[String]()
+
+  def this(loggers: Array[Logger], settings: RunSettings,
+      testClassName: String) = {
+    this(loggers, settings)
+    currentTestClassName.push(testClassName)
+  }
+
+  def pushCurrentTestClassName(s: String): Unit =
+    currentTestClassName.push(s)
+
+  def popCurrentTestClassName(): Unit =
+    if (currentTestClassName.size > 1) currentTestClassName.pop()
+
+  def debug(s: String): Unit = {
+    for (l <- loggers) {
+      if (settings.color && l.ansiCodesSupported()) l.debug(s)
+      else l.debug(filterAnsi(s))
+    }
+  }
+
+  def error(s: String): Unit = {
+    for (l <- loggers) {
+      if (settings.color && l.ansiCodesSupported()) l.error(s)
+      else l.error(filterAnsi(s))
+    }
+  }
+
+  def error(s: String, t: Throwable): Unit = {
+    error(s)
+    if (t != null && (settings.logAssert || !t.isInstanceOf[AssertionError]))
+      logStackTrace(t)
+  }
+
+  def info(s: String): Unit = {
+    for (l <- loggers) {
+      if (settings.color && l.ansiCodesSupported()) l.info(s)
+      else l.info(filterAnsi(s))
+    }
+  }
+
+  def warn(s: String): Unit = {
+    for (l <- loggers) {
+      if (settings.color && l.ansiCodesSupported()) l.warn(s)
+      else l.warn(filterAnsi(s))
+    }
+  }
+
+  private def logStackTrace(t: Throwable): Unit = {
```

Comment:
How well does all this stack trace manipulation work on the various JS engines? With and without source maps? This is notoriously very fragile, and we need to make sure we don't completely destroy a stack trace that is only partially compliant with what the JVM gives.